packages/flutter/lib/src/widgets/modal_button_sheet.dart

```diff
@@ -0,0 +1,652 @@
+                    ),
+                  ),
+                ),
+              );
+        return ClipRect(
+          child: CustomSingleChildLayout(
+            delegate: _ModalBottomSheetLayout(
+              animationValue,
+              widget.expanded,
+            ),
+            child: draggableChild,
+          ),
+        );
+      },
+    );
+
+    return ScrollToTopStatusBarHandler(
+      child: child,
+      scrollController: _scrollController,
+    );
+  }
+}
+
+class _ModalBottomSheetLayout extends SingleChildLayoutDelegate {
+  _ModalBottomSheetLayout(this.progress, this.expand);
+
+  final double progress;
+  final bool expand;
+
+  @override
+  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
+    return BoxConstraints(
+      minWidth: constraints.maxWidth,
+      maxWidth: constraints.maxWidth,
+      minHeight: expand ? constraints.maxHeight : 0,
+      maxHeight: expand ? constraints.maxHeight : constraints.minHeight,
+    );
+  }
+
+  @override
+  Offset getPositionForChild(Size size, Size childSize) {
+    return Offset(0.0, size.height - childSize.height * progress);
+  }
+
+  @override
+  bool shouldRelayout(_ModalBottomSheetLayout oldDelegate) {
+    return progress != oldDelegate.progress;
+  }
+}
+
+class _CustomBottomSheetLayout extends SingleChildLayoutDelegate {
+  _CustomBottomSheetLayout(this.progress);
+
+  final double progress;
+  double? childHeight;
+
+  @override
+  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
+    return BoxConstraints(
+      minWidth: constraints.maxWidth,
+      maxWidth: constraints.maxWidth,
+      minHeight: constraints.minHeight,
+      maxHeight: constraints.maxHeight + progress * 8,
+    );
+  }
+
+  @override
+  Offset getPositionForChild(Size size, Size childSize) {
+    childHeight ??= childSize.height;
+    return Offset(0.0, size.height - childSize.height);
+  }
+
+  @override
+  bool shouldRelayout(_CustomBottomSheetLayout oldDelegate) {
+    if (progress != oldDelegate.progress) {
+      childHeight = oldDelegate.childHeight;
+      return true;
+    }
+    return false;
+  }
+}
+
+// Checks the device input type as per the OS installed in it
+// Mobile platforms will be default to `touch` while desktop will do to `mouse`
+// Used with VelocityTracker
+// https://github.com/flutter/flutter/pull/64267#issuecomment-694196304
+PointerDeviceKind defaultPointerDeviceKind(BuildContext context) {
+  final platform =  defaultTargetPlatform;
+  switch (platform) {
+    case TargetPlatform.iOS:
+    case TargetPlatform.android:
+      return PointerDeviceKind.touch;
+    case TargetPlatform.linux:
+    case TargetPlatform.macOS:
+    case TargetPlatform.windows:
+      return PointerDeviceKind.mouse;
+    case TargetPlatform.fuchsia:
+      return PointerDeviceKind.unknown;
+  }
+  return PointerDeviceKind.unknown;
+}
+
+// Copied from bottom_sheet.dart as is a private class
+// https://github.com/flutter/flutter/issues/51627
```

Comment:
This class `BottomSheetSuspendedCurve` I think should be made public, moved to another file, and be reused as it would be now three times duplicated.